Use QuickOpen to load resources in the inspector.

rcorre opened the pull request:
When you click the "Load" button to populate a Resource field in the
inspector dock, the editor currently creates a file dialog. Navigating
through folders to find the file you want is tedious.

This replaces the file dialog with the same QuickOpen dialog used for
the "Quick Run" and "Quick Open" scene actions. The result list is
filtered to only resources of the appropriate type, and you can type
I find that this is much faster and more intuitive than the file dialog.

As far as I'm aware, the only "feature" you lose compared to the old
FileDialog is the ability to show all types of files by changing the
extension filter to "All Files" in the lower right. This seems like an
unnecessary feature anyways, as selecting a file that is not of the
appropriate resource type would just result in an error like so:

```
The selected resource (StreamTexture) does not match any type expected for this property (AudioStream).
```

Relates to godotengine/godot-proposals#346.

The old way:
![filedialog](https://user-images.githubusercontent.com/2496231/77250922-96ec2380-6c21-11ea-8566-8e4cab1e0dd4.gif)

The new way:
![quickopen](https://user-images.githubusercontent.com/2496231/77250923-9784ba00-6c21-11ea-8c82-6e39f6d7d19b.gif)

rcorre committed:
Add QuickLoad option to resource picker.

When clicking on a resource field in the inspector dock, you now have
the "Quick Load" option in addition to "Load". This opens a QuickOpen
dialog allowing the user to type in a phrase to quickly locate the
desired resource (similar to "Quick Open Scene").

In my experience, this is much faster than clicking through the File
Dialog.

Relates to godotengine/godot-proposals#346.

rcorre commented:
I should probably call `set_title` on the `QuickOpen` to say something like "Select a Resource" instead of "Please Confirm". I figure that should be wrapped in a `TTR`, but what else do I need to do so it gets translated?

Calinou commented:
> I figure that should be wrapped in a TTR, but what else do I need to do so it gets translated?

Wrapping it in a `TTR()` call should be enough.

rcorre commented:
Rebased and squashed. FWIW I'm also playing with using quickopen for the "New Inherited Scene" command at https://github.com/rcorre/godot/commit/55e953ace5393f782d366d68f6a92106f9a28b73 but haven't added it to this PR.

akien-mga commented:
> As far as I'm aware, the only "feature" you lose compared to the old
FileDialog is the ability to show all types of files by changing the
extension filter to "All Files" in the lower right.

akien-mga commented:
> As far as I'm aware, the only "feature" you lose compared to the old
FileDialog is the ability to show all types of files by changing the
extension filter to "All Files" in the lower right.

You also lose the resource preview that the FileDialog shows. In your GIF it's arguably not the most important as they're sound waves, but for images or scenes, the preview can be quite useful to pick the relevant asset. With QuickOpen you need to know what's the exact name of the resource you want to load, without any option to preview it.

rcorre commented:
I guess I tend to focus on names rather than images, but fair point. Do you think it is feasible/useful to add a preview to quickopen similar to [fzf's preview window](https://github.com/junegunn/fzf#preview-window)?

Calinou commented:
Given the current state of the QuickOpen dialog, this would likely be better as a opt-in editor setting for power users.

rcorre commented:
I'd be happy with this as an opt-in. I can also look into adding preview if that seems plausible -- if we can make this work for all cases and avoid having two code paths that seems ideal.

I use this feature all the time on my personal editor build, so I'd love to see it in the mainline.

pycbouh commented:
This looks interesting, but if it's still desired it needs to be remade for the current EditorResourcePicker control. To address the concerns voiced by others I would suggest adding another menu option besides "Load", "Quick Load" that would bring up that dialog. No editor setting needed, both options always handy.

rcorre commented:
Thanks @pycbouh, I've updated it to 4.0 and made it a separate option from the regular load dialog.
In 4.0, I've found that the QuickOpen dialog does not automatically grab focus, which makes it much less useful (you still have to grab your mouse and mouse over it). That happens in the "Quick Open Scene" dialog as well though, so I don't think it is a problem with this PR.

pycbouh commented:
> In 4.0, I've found that the QuickOpen dialog does not automatically grab focus, which makes it much less useful (you still have to grab your mouse and mouse over it).

Hmm, it grabs focus for me. Can be a platform-specific problem due to dialogs being real windows now.

YuriSizov reviewed (approved):
It works well, I like the feature and that it's side by side with the regular "Load" now. Code also looks good and simple, easy to maintain in future. I would probably go for a new icon here, maybe the same folder with a looking glass superimposed, but it can be done in another PR.

You just need to fix the include order to make clang-format happy, and should be good to merge (if there are no objections).


Here's how it currently works:

https://user-images.githubusercontent.com/11782833/133436426-fe91a32c-7c65-4c06-8419-108d08b4a448.mp4

Calinou commented:
Should Quick Load be listed *before* Load in the menu? I think most people will prefer to use Quick Load as a "default" option, and only use Load when they actually need thumbnails.

pycbouh commented:
> Should Quick Load be listed _before_ Load in the dropdown? I think most people will prefer to use Quick Load as a "default" option, and only use Load when they actually need to see thumbnails.

Semantically it's probably reasonable to have it before, but from the usability perspective I don't see much difference as they are next to each other and it's a matter of a few pixels.

rcorre commented:
> > In 4.0, I've found that the QuickOpen dialog does not automatically grab focus, which makes it much less useful (you still have to grab your mouse and mouse over it).
> 
> Hmm, it grabs focus for me. Can be a platform-specific problem due to dialogs being real windows now. But it is indeed out of scope here, if that's the case.

Ah, "single window" got unset for me somehow, and I use focus-follows-mouse in my window manager.

pycbouh commented:
> Ah, "single window" got unset for me somehow, and I use focus-follows-mouse in my window manager.

👍 As long as it's not on Godot 🙂 

-----

Can you reverse the order of the options so that they make more sense semantically (quick before regular)?

rcorre commented:
> Can you reverse the order of the options so that they make more sense semantically (quick before regular)?

Done

akien-mga commented:
Thanks!

akien-mga commented:
Cherry-picked for 3.4.
